SpatialGDK/Source/SpatialGDK/Private/SpatialView/Dispatcher.cpp

```diff
@@ -0,0 +1,230 @@
+﻿// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "SpatialView/Dispatcher.h"
+
+#include "Algo/BinarySearch.h"
+#include "SpatialView/EntityComponentTypes.h"
+
+namespace SpatialGDK
+{
+FDispatcher::FDispatcher(const EntityView* View)
+	: View(View)
+	, NextCallbackId(1)
+{
+}
+
+void FDispatcher::InvokeCallbacks(const TArray<EntityDelta>& Deltas)
+{
+	for (const auto& EntityDelta : Deltas)
+	{
+		HandleComponentPresenceChanges(EntityDelta.EntityId, EntityDelta.ComponentsAdded, &FComponentCallbacks::ComponentAddedCallbacks);
+		HandleComponentPresenceChanges(EntityDelta.EntityId, EntityDelta.ComponentsRemoved,
+									   &FComponentCallbacks::ComponentRemovedCallbacks);
+		HandleComponentValueChanges(EntityDelta.EntityId, EntityDelta.ComponentUpdates);
+		HandleComponentValueChanges(EntityDelta.EntityId, EntityDelta.ComponentsRefreshed);
+
+		HandleAuthorityChange(EntityDelta.EntityId, EntityDelta.AuthorityGained, &FAuthorityCallbacks::AuthorityGainedCallbacks);
+		HandleAuthorityChange(EntityDelta.EntityId, EntityDelta.AuthorityLost, &FAuthorityCallbacks::AuthorityLostCallbacks);
+		HandleAuthorityChange(EntityDelta.EntityId, EntityDelta.AuthorityLostTemporarily,
+							  &FAuthorityCallbacks::AuthorityLostTemporarilyCallbacks);
+	}
+}
+
+CallbackId FDispatcher::RegisterComponentAddedCallback(Worker_ComponentId ComponentId, FComponentValueCallback Callback)
+{
+	InvokeWithExistingValues(ComponentId, Callback);
+	const int32 Index = Algo::LowerBound(ComponentCallbacks, ComponentId, FComponentCallbacks::ComponentIdComparator());
+	if (Index == ComponentCallbacks.Num() || ComponentCallbacks[Index].Id != ComponentId)
+	{
+		ComponentCallbacks.EmplaceAt(Index, ComponentId);
+	}
+	ComponentCallbacks[Index].ComponentAddedCallbacks.Register(NextCallbackId, MoveTemp(Callback));
+	return NextCallbackId++;
+}
+
+CallbackId FDispatcher::RegisterComponentRemovedCallback(Worker_ComponentId ComponentId, FComponentValueCallback Callback)
+{
+	const int32 Index = Algo::LowerBound(ComponentCallbacks, ComponentId, FComponentCallbacks::ComponentIdComparator());
+	if (Index == ComponentCallbacks.Num() || ComponentCallbacks[Index].Id != ComponentId)
+	{
+		ComponentCallbacks.EmplaceAt(Index, ComponentId);
+	}
+	ComponentCallbacks[Index].ComponentRemovedCallbacks.Register(NextCallbackId, MoveTemp(Callback));
+	return NextCallbackId++;
+}
+
+CallbackId FDispatcher::RegisterComponentValueCallback(Worker_ComponentId ComponentId, FComponentValueCallback Callback)
+{
+	InvokeWithExistingValues(ComponentId, Callback);
```

Comment:
nit: this seems opinionated, but could always be what a user wants. Could it make sense to offer two register callback functions for each event type, one which invokes if it should be and then registers, and one that just registers? This could also be added later if it turns out we need it I guess.

Reply:
discussed offline - this is a first guess based on the prototypes but we can start with explicit Register and RegisterAndInvoke functions that will give a first callback if appropriate.